frontend/ui/widget/inputtext.lua

```diff
@@ -139,7 +139,7 @@ if Device:isTouchDevice() or Device:hasDPad() then
                     self.is_keyboard_hidden = false
                 end
             end
-            if #self.charlist > 0 then -- Avoid cursor moving within a hint.
+            if #self.charlist > 0 and self._frame_textwidget.dimen then -- Avoid cursor moving within a hint.
```

Comment:
Not sure why you had to / need to add this condition - as it feels it should always exists ?
Is this a requirement for how this new keyboard works - or just a bug and something we forgot that you logically fixed.
Please add a comment if the former (saying it's for the zh keyboard).

Reply:
I have it that when tapping on the input text box, I need to remove the candidates if they exist. Then when I'm modifying the text, the dimen here somehow becomes `nil` then crash. I remember in Ja keyboard it happened too, but I couldn't make it happen again. With the Chinese keyboard, with candidates shown, it always appears.

Reply:
See ! How readable your above comment is. Imagine I had to read:
> I need to remove the candi if they exist. [...] With the Chinese keyboard, with candi shown, it always appears.

Your code should read as easy as your posts :)
(Why had you removed the sweets ?! :)

Reply:
I'm not familiar with the widget in question, so this is more of a generic comment about this specific pitfall:

`dimen` fields are often set at *draw* time (e.g., in `paintTo`), so for widgets that get destroyed, this may not always be set when you expect it to.

Depending on what makes the most sense, you can either side-step the issue like you did, fix the widget so that stuff doesn't get randomly destroyed if it doesn't make much sense, or forcibly compute dimen before paint via the `getSize` method ;).

Reply:
One practical consequence of this is the difference between the two signatures for `setDirty`: when you pass a lambda, it's often because you want to use a `dimen` field for the refresh region, but it might not be set yet at call time. But since UIManager consumes the lambda post-paintTo, it magically works ;o).

Reply:
I tried     
```
if not inputbox._frame_textwidget.dimen then
    inputbox._frame_textwidget.dimen = inputbox._frame_textwidget:getSize()
end
```
after calling `inputbox:delChar`, it's no longer crashing :) But somehow I still can't immediately have the cursor where I clicked. It may be because the new text is not drawn yet? 
Also I'm here modifying the supposedly private property of inputbox thanks to Lua letting me do it. So I think I stick with the side-step, as my case is quite rare.

Reply:
Yeah, depending on what exactly you call getSize() on, it may not always behave as expected (e.g., a container with multiple child widgets, some of 'em being undrawn/destroyed or whatever), so, it can indeed be a bit iffy ;).

(I think there's still an insane bit of hackery for a similar shenanigan in ReaderFooter that still gives me nightmares ;p).